### PR TITLE
Fix spell check bug

### PR DESCRIPTION
**Pull request by zhujinxuan**
I find there is a solution in draft.js by https://github.com/facebook/draft-js/blob/master/src/component/contents/DraftEditorTextNode.react.js#L36

which is to add `data-*` in each layer of marks

This PR fixes the slate spellcheck problem in the same way.

After this fix, renderMark should have props.attributes to append `data-text:true`

Cheers,
Jinxuan Zhu

### CODE REVIEW

File: packages/slate-react/src/plugins/after.js [hunk 1]
```diff
@@ -305,8 +305,8 @@ function AfterPlugin() {
 
     // Get the selection point.
     const native = window.getSelection()
-    const { anchorNode, anchorOffset } = native
-    const point = findPoint(anchorNode, anchorOffset, value)
+    const { anchorNode } = native
+    const point = findPoint(anchorNode, 0, value)
```
Comment (ianstormtaylor) [inline packages/slate-react/src/plugins/after.js]: Why this change?

File: packages/slate-react/src/plugins/after.js [hunk 2]
```diff
@@ -323,7 +323,7 @@ function AfterPlugin() {
       leaves.find(r => {
         start = end
         end += r.text.length
-        if (end >= point.offset) return true
+        if (end > point.offset) return true
```
Comment (ianstormtaylor) [inline packages/slate-react/src/plugins/after.js]: Why this change?

File: packages/slate-react/src/components/leaf.js [hunk 3]
```diff
@@ -86,7 +90,11 @@ class Leaf extends React.Component {
       index,
     })
 
-    return <span data-offset-key={offsetKey}>{this.renderMarks()}</span>
+    return (
+      <span ref={this.setRef} data-offset-key={offsetKey}>
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/leaf.js]: Why this `ref=`?

File: packages/slate-react/src/components/leaf.js [hunk 4]
```diff
@@ -101,7 +109,16 @@ class Leaf extends React.Component {
...
+      const props = {
+        editor,
+        mark,
+        marks,
+        node,
+        offset,
+        text,
+        children,
+        attributes,
+      }
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/leaf.js]: Can you just define `attributes` inline here instead of pulling it up above, since it's only used in one place?

File: packages/slate-react/src/components/leaf.js [hunk 5]
```diff
@@ -5,6 +5,10 @@ import SlateTypes from 'slate-prop-types'
 
 import OffsetKey from '../utils/offset-key'
 
+const attributes = {
+  'data-text': true,
+}
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/leaf.js]: Please see comment below to remove this.

File: packages/slate-react/src/components/leaf.js [hunk 6]
```diff
@@ -5,6 +5,10 @@ import SlateTypes from 'slate-prop-types'
 
 import OffsetKey from '../utils/offset-key'
 
+const attributes = {
+  'data-text': true,
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/leaf.js]: Can you rename this to `data-slate-leaf: true` to match the other data attributes we've got?

File: examples/rich-text/index.js [hunk 7]
```diff
@@ -302,16 +302,16 @@ class RichTextExample extends React.Component {
...
+        return <strong {...attributes}>{children}</strong>
       case 'code':
-        return <code>{children}</code>
+        return <code {...attributes}>{children}</code>
       case 'italic':
-        return <em>{children}</em>
+        return <em {...attributes}>{children}</em>
       case 'underlined':
-        return <u>{children}</u>
+        return <u {...attributes}>{children}</u>
```
Comment (ianstormtaylor) [inline examples/rich-text/index.js]: Can you update the other examples that use `renderMark` so that all of them mix in the `attributes` now?

File: packages/slate-react/src/plugins/after.js (hunk 1 above)
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: Consider a leaf can be extended by a spell check:
```
<b>hel</b> yourself
```
will be 
```
<b>help</b> yourself
```
after spell check.  Because `anchorOffset` is measured after the spell check.  It will tell that anchorOffset is `4`.  However, during `onInput`, the text is not yet changed.  `offset =4` will tell the slate the spell corrected leaf is `yourself`, not `hel`

File: packages/slate-react/src/plugins/after.js (hunk 2 above)
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: The same reason of the previous explanation:
Consider a leaf can be extended by a spell check:
```
<b>hel</b> yourself
```
will be
```
<b>help</b> yourself
```
after spell check. Because anchorOffset is measured after the spell check. It will tell that anchorOffset is 4. However, during onInput, the text is not yet changed. offset =4 will tell the slate the spell corrected leaf is yourself, not hel

File: packages/slate-react/src/components/leaf.js (hunk 3 above)
Comment (zhujinxuan) [reply packages/slate-react/src/components/leaf.js]: I forgot to remove it. I will remove that.

File: examples/rich-text/index.js (hunk 7 above)
Comment (zhujinxuan) [reply examples/rich-text/index.js]: Fixed~

File: packages/slate-react/src/components/leaf.js (hunk 4 above)
Comment (zhujinxuan) [reply packages/slate-react/src/components/leaf.js]: Fixed, thank you

Comment (zhujinxuan) [reply packages/slate-react/src/components/leaf.js]: If we want to define attributes in `reduce` function; how about `data-slate-mark:index`; then we may be able to find leaf-mark by index in future development if necessary.

File: packages/slate-react/src/components/leaf.js (hunk 5 above)
Comment (zhujinxuan) [reply packages/slate-react/src/components/leaf.js]: Fixed~

File: packages/slate-react/src/components/leaf.js (hunk 6 above)
Comment (zhujinxuan) [reply packages/slate-react/src/components/leaf.js]: Fixed~

**Comment by zhujinxuan** [comment]
Could I know if there is any updates?

**Comment by ianstormtaylor** [comment]
Thanks @zhujinxuan, great find here.
